### PR TITLE
String → Identifier

### LINKED ISSUE #227: [RFC] Typed identifiers
I noticed NOVA always uses string identifiers. IMO NOVA should use typed identifiers:

1. `IIdentifier` interface
    
    - The `I` at the beginning is necessary to avoid conflicts with a bunch of things (e.g. Kotlin). Could also be `NOVAIdentifier` - it just can't be `Identifier`.
    - Single method: `String asString();`
    - Implementations should override `equals`, `hashCode` and `toString`.

2. `UUIDIdentifier` class
    
    - Used when the identifier is an `UUID`.
    - Additional methods: `UUID asUUID();`

3. `ClassIdentifier` class
    
    - Used when the identifier is a `Class`.
    - Additional methods: `Class<?> asClass();`

4. `StringIdentifier` class
    
    - Used when the identifier is an actual, proper `String`.

Will make a PR once we figure out if this is a good idea.

### PR DESCRIPTION
**Pull request by ExE-Boss**
@SoniEx2 has pretty much abandoned his PR, so I’m taking over.
For a detailed description of what this does, see #227.

### To do list:
- [ ] Get rid of `String` serialization.
- [ ] Redesign this whole thing from scratch. (Seriously, the current implementation is very bloated)
    - [ ] Get rid of `IdentifierFactory` (or whatever I called it) and replace it with the `Storable` interface.

### Completed:
- [x] Implement `StringIdentifier`, `ClassIdentifier` and `UUIDIdentifier` (SoniEx2)
- [x] Implement `NamespacedStringIdentifier` and `EnumIdentifier`
    - IDs can’t be `null`. That throws a `NullPointerException`.
- [x] Switch Core to use `Identifier` everywhere
    - Currently kept an implicit `String` → `StringIdentifier` method in `FactoryManager`
      Should we keep it or not?
- [x] Write tests
- [x] Fix wrappers
    - [x] Test in Minecraft 1.7 and 1.8

---

Closes #227

### CODE REVIEW

**Comment by SoniEx2** [comment]
`null` IDs?!

**Comment by SoniEx2** [comment]
You need to learn how to use a damn immutable object.

And a constructor for such an immutable object.

File: src/main/java/nova/core/util/id/AbstractIdentifier.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+	@Override
+	public int hashCode() {
+		return Objects.hashCode(id);
+	}
+
+	@Override
+	public abstract boolean equals(Object other);
+
+	protected static final <T extends Identifier> boolean equalsImpl(Identifier _this, Object other, Class<T> superclass, Function<T,Object> getter) {
```
Comment (RX14) [inline src/main/java/nova/core/util/id/AbstractIdentifier.java]: `self` instead of `_this`? Is there a convention on this like `clazz`?

**Comment by Victorious3** [comment]
Ive yet to understand why you'd need these. Interned strings are enough of an identifier to me...

File: src/test/java/nova/core/util/EnumExample.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+package nova.core.util;
+
+/**
+ * @author ExE Boss
+ */
+public enum EnumExample {
```
Comment (RX14) [inline src/test/java/nova/core/util/EnumExample.java]: This change shouldn't be in this PR.

**Comment by RX14** [comment]
@Victorious3 wouldn't interned strings technically work with `==`? I wouldn't recommend it because it might be a bit flakey but if they are interned shouldn't they be the "same" string.

File: src/test/java/nova/core/util/EnumExample.java (hunk 2 above)
Comment (ExE-Boss) [inline src/test/java/nova/core/util/EnumExample.java]: This change was done to allow testing of `EnumIdentifier` using my `EnumExample` class.

Comment (RX14) [inline src/test/java/nova/core/util/EnumExample.java]: Oh, okay.

File: src/main/java/nova/core/util/id/AbstractIdentifier.java (hunk 1 above)
Comment (ExE-Boss) [inline src/main/java/nova/core/util/id/AbstractIdentifier.java]: Done now.

**Comment by Victorious3** [comment]
@RX14 Yes they do see my 0815 implementation. And 
```java
Identifier.of("foo") == Identifier.of(new String(new char[] {'f', 'o', 'o'})));
```

**Comment by SoniEx2** [comment]
@Victorious3 The issue is Identifier.of("net.minecraft.block.Block") == Identifier.of(net.minecraft.block.Block.class). That should NEVER be able to happen.

**Comment by RX14** [comment]
@SoniEx2 How about you calm the fuck down and suggest improvements nicely instead of shouting at people. It won't make them take your input more seriously.

**Comment by SoniEx2** [comment]
@RX14 I'll happily shout at someone who's touching code I wrote without any idea of what the code should do.

**Comment by RX14** [comment]
@SoniEx2 Well, if you keep doing it you won't be commenting on this issue tracker any more. I suggest you talk nicely or don't talk at all.

**Comment by SoniEx2** [comment]
@RX14 Fine.

Part of the reason I made the PR was so I could put different kinds of identifiers in the same HashMap without having them conflict.

A NamespacedStringIdentifier should contain an inner static class called NamespacedStringData, and that class should be used as the identifier type. The equals and hashCode methods should be final, since otherwise equals wouldn't be symmetric. This is also why we can't/shouldn't have null IDs.

**Comment by SoniEx2** [comment]
Also isn't NOVA supposed to work in pre-1.7.10 versions of minecraft, aka blockID versions of minecraft?

**Comment by RX14** [comment]
Mapping different types of identifiers to strings in a non-conflicting way should be possible: just add a prefix.

**Comment by SoniEx2** [comment]
@RX14 That's one way to make it unmaintainable. How would you handle `Identifier<Integer>` as a string?

"i:1"?

"i:\u0001"?

**Comment by Victorious3** [comment]
@SoniEx2 
`Identifier.of("net.minecraft.block.Block") == Identifier.of(net.minecraft.block.Block.class)`
Sure. This is fine tho. Where's the problem with that?

**Comment by SoniEx2** [comment]
@Victorious3 One's a class, the other is a string. The `Identifier` class is a *wrapper* class to allow type-safe usage of identifier types in objects. (We could just use Object instead, which would've avoided this whole argument, but that's not type-safe.)

In other words, just like "net.minecraft.block.Block" != net.minecraft.block.Block.class, Identifier.of("net.minecraft.block.Block") shouldn't be equal to Identifier.of(net.minecraft.block.Block.class).

**Comment by Victorious3** [comment]
@SoniEx2 Typesafe identifiers? Please? Are you even listening to yourself? The purpose of an identifier is to identify any given object in a specific *context*. If its about identifying Blocks you'll identify blocks and nothing else. It doesn't matter if an item and a block, or an entity share the same id because they are stored in different places and you are using different methods to access them. An identifier doesn't have to be unique. It CAN be unique, but thats by all means no requirement.

**Comment by RX14** [comment]
I can imagine identifiers being unique aiding debuggability.

**Comment by SoniEx2** [comment]
Old versions of Minecraft use ints instead of strings. To support both in a highly compatible way, you use Identifier class, with StringIdentifier/NamespacedStringIdentifier and IntegerIdentifier.

**Comment by Victorious3** [comment]
@RX14 Well then use UUIDs everywhere and call your blocks "d6356985-e8bd-45ab-b16c-8a589af95cc9" instead of stone, gravel or whatever. This PR is overcomlicating a non issue.

**Comment by SoniEx2** [comment]
Fine. Replace all instances of "Identifier" with "Object". That'll solve all your issues.

**Comment by Victorious3** [comment]
@SoniEx2 You mean String. Like it was before.

**Comment by SoniEx2** [comment]
@Victorious3 I mean Object, for its higher flexibility.

**Comment by Victorious3** [comment]
No, because you can't serialize "Object" so its useless as an identifier.

**Comment by SoniEx2** [comment]
And string is useless for older Minecraft versions.

**Comment by Victorious3** [comment]
@SoniEx2 This is Nova CORE it doesn't CARE about "Minecraft Versions"

**Comment by SoniEx2** [comment]
And the Minecraft wrappers are directly integrated into NOVA Core (kinda why they're in the same repo instead of being in their own repos).

**Comment by Victorious3** [comment]
@SoniEx2 If you manage to name one problem that this overbloated PR supposedly solves, then go for it. You are shooting with cannons at ants. I'm for simplicity and ease of use and things like `NamespacedStringIdentifier` certainly don`t fall into that category.

**Comment by SoniEx2** [comment]
Please make Mojang remove Minecraft's ResourceLocation. It's bloated.

**Comment by Victorious3** [comment]
@SoniEx2 Besides this being completely off point, ResourceLocation is NOT an identifier. It serves the purpose of being able to locate files within the jar or resource packs which use completely different file paths.

**Comment by SoniEx2** [comment]
@Victorious3 It is actually NO different from a NamespacedStringIdentifier. It identifies block names, item names, etc.

**Comment by Victorious3** [comment]
@SoniEx2 You are impossible to talk to. Closed until further notice from @ExE-Boss

**Comment by RX14** [comment]
@Victorious3 While I share your opinion, that doesn't mean I won't remove you from the project or issue tracker if you continue to be rude.

**Comment by SoniEx2** [comment]
You just need to talk to some java pros about it. Maybe even @calclavia (altho I'm not sure how "pro" they are).

**Comment by ExE-Boss** [comment]
On the NOVA discord server, I’ve discussed this with RX14 and we decided the following:

I would continue working on my version of the Identifier AP and once all the versions of the Identifier API are fully implemented, we will compare them and decide which implementation is the best and end up using that.

**Comment by SoniEx2** [comment]
I *need* something I can use with HashMaps and stuff. I *cannot* use this implementation, and strings are unsuitable due to their conflicting nature.

**Comment by Victorious3** [comment]
@SoniEx2 You dont throw two different types of Objects into the same map. Just dont do it. Then you are fine.

**Comment by SoniEx2** [comment]
I can throw String and Class into the same Map just fine. They don't conflict with eachother.

I want it to also work for Identifiers.

**Comment by Victorious3** [comment]
@SoniEx2 And then its a `Map<Object, Object>` hell no. 
>Like using Identifier for a programming language/compiler.

Mhm. Sounds like something I'd do every day.

**Comment by loordgek** [comment]
@Victorious3 ResourceLocation IS used as a identifier https://github.com/MinecraftForge/MinecraftForge/blob/1.11.x/src/main/java/net/minecraftforge/event/AttachCapabilitiesEvent.java#L74
https://github.com/MinecraftForge/MinecraftForge/blob/1.11.x/src/main/java/net/minecraftforge/fml/common/registry/IForgeRegistryEntry.java#L50

**Comment by Victorious3** [comment]
@loordgek And so might any other object that implements `hashcode` and `equals`. Yet you wouldn't call them identifiers.

**Comment by SoniEx2** [comment]
I miss @calclavia, the only professional/sane java dev we had on this POS.

**Comment by loordgek** [comment]
@SoniEx2 can you stop crying ranting plz

**Comment by calclavia** [comment]
@ExE-Boss What's the NOVA discord server? I'd like to join just to see what you guys are up to.

**Comment by ExE-Boss** [comment]
@calclavia The server is https://discord.gg/Sh63S4H

**Comment by SoniEx2** [comment]
@loordgek Never. If you don't like it, why do you keep doing things that cause it?

**Comment by RX14** [comment]
@SoniEx2 goodbye

**Pull request closed without merging**